File: OP_PROJECT_JAVA/src/game/Animal.java
package game;

import java.util.Objects;
import java.util.Random;

public class Animal extends Organism {
    public Animal(String type, int strength, int initiative, Position position, World world) {
        super(type, strength, initiative, position, world);
    }

    @Override
    public void action() {
        Random random = new Random();
        int direction = random.nextInt(4);
        Position newPosition = new Position(position.getY(), position.getX());

        settingDirection(direction, newPosition);

        if (world.isPositionWithinBounds(newPosition)) {
            if (world.getOrganismAtPosition(newPosition) != null) {
                collision(world.getOrganismAtPosition(newPosition));
            } else {
                moveOrganism(newPosition);
            }
        }
    }

    public void moveOrganism(Position newPosition) {
        world.removeOrganismAtPosition(position);
        position = newPosition;
        world.addOrganismAtPosition(position, this);
    }

    @Override
    protected boolean checkBorn(Organism first, Organism second) {
        if (Objects.equals(first.getType(), second.getType())) {
            Position lPosition;
            do {
                lPosition = new Position((int) (Math.random() * world.getHeight()), (int) (Math.random() * world.getWidth()));
            } while (world.isPositionEmpty(lPosition));
            world.bornAnimal(first.getType(), lPosition);
            return true;
        }
        return false;
    }

    @Override
    public OrganismGroup getGroup() {
        return OrganismGroup.ANIMAL;
    }
}

File: OP_PROJECT_PYTHON/src/game/Animal.py
from random import randint
from Organism import Organism
from Position import Position


class Animal(Organism):
    def __init__(self, organism_type, strength, initiative, position, world):
        super().__init__(organism_type, strength, initiative, position, world)

    def action(self):
        direction = randint(0, 3)
        new_position = Position(self.position.get_y(), self.position.get_x())
        self.setting_direction(direction, new_position)

        if self.world.is_position_within_bounds(new_position):
            organism_at_new_position = self.world.get_organism_at_position(new_position)
            if organism_at_new_position is not None:
                self.collision(organism_at_new_position)
            else:
                (self.move(new_position))

    def check_born(self, first, second):
        if first.get_type() == second.get_type():
            while True:
                new_position = Position(randint(0, self.world.height - 1), randint(0, self.world.width - 1))
                if self.world.is_position_empty(new_position):
                    self.world.born_organism(first.get_type(), new_position)
                    return True
        return False

    def get_group(self):
        return "ANIMAL"
